### PR TITLE
FIP-0054: Filecoin EVM runtime (FEVM)

### PR DESCRIPTION
**Pull request by raulk**
## Simple Summary

We introduce a new built-in actor: the Filecoin EVM (FEVM) runtime actor, capable of running EVM smart contracts on top of the Filecoin Virtual Machine. We also introduce various changes to the Filecoin Virtual Machine and to client implementations, necessary to support the operation of this new actor.

## Abstract

The Filecoin EVM (FEVM) runtime built-in actor runs EVM smart contracts compatible with the [Ethereum Paris fork].

It achieves this by embedding a special-purpose EVM interpreter, implementing the integration logic with the Filecoin environment, translating environment-dependent opcodes into their corresponding Filecoin primitives, and mapping all state I/O to the underlying IPLD model.

The EVM interpreter strives for maximal as-is portability. To such avail, it The EVM interpreter supports all required opcodes and Ethereum precompiles to strive for maximal portability. Functional and technical departures from Ethereum's standard expectations are documented herein.

The FEVM runtime actor motivates some changes in the FVM. Syscalls are modified, syscalls are added, new environmental data is required, and new exit codes are created. The concept of a TipsetCID is also introduced formally, and is of required implementation by clients.

This FIP is dependent on [FIP-0048] (f4 address class), [FIP-0049] (Actor events), and [FIP-0055] (Supporting Ethereum Accounts, Addresses, and Transactions).

[Ethereum Paris fork]: https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/
[FIP-0048]: https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0048.md
[FIP-0049]: https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0049.md
[FIP-0055]: https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0055.md

### CODE REVIEW

File: FIPS/fip-nnnn-evm-compatibility.md [hunk 1]
```diff
@@ -0,0 +1,1079 @@
+---
+fip: <to be assigned>
```
Comment (jennijuju) [inline FIPS/fip-nnnn-evm-compatibility.md:2]: 0053

File: FIPS/fip-nnnn-eam-eeoa.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+We also introduce **Ethereum addresses** atop the f4 address class, and a new **Delegated signature type**.
+
+## Abstract
+
+The **Ethereum Address Manager (EAM)** actor manages namespace `10` within the `f4` address class as defined in [FIP-0048].
+This namespace hosts **Ethereum addresses** in Filecoin.
+These addresses begin with `f410` in textual form.
+EVM smart contracts deployed through the EAM acquire an `f410` address.
+
+The **Ethereum Externally Owned Account (EEOA)** actor represents Ethereum identities backed by secp256k1 keys.
```
Comment (jennijuju) [inline FIPS/fip-nnnn-eam-eeoa.md]: i thought we agreed on EthAccount (no more EEOA)?

Comment (raulk) [reply FIPS/fip-nnnn-eam-eeoa.md]: I learnt about the rename recently, and I agree with the motivation. It is already updated, but I think you added this comment before I pushed my latest changes.

File: FIPS/fip-0054.md [hunk 3]
```diff
@@ -399,7 +399,26 @@ attoFIL (same precision as Ethereum).
...
+Each variant takes the specified number of topics (indexed).
+All variants take a single memory slice referring to the data (non indexed).
+
+We define `LOG{0..4}` opcodes to emit [FIP-0049] compliant actor events conforming to the following template:
+
+```rust
+// Values are encoded as DAG-CBOR byte strings.
+ActorEvent {
+    entries: [
+        (0x03, "topic1", <first topic word>),   // when LOG1, LOG2, LOG3, LOG4
```
Comment (maciejwitowski) [inline FIPS/fip-0054.md:414]: @raulk it looks like in code all entries have flag `0x02` ([code](https://github.com/filecoin-project/builtin-actors/blob/next/actors/evm/src/interpreter/instructions/log.rs#L42). Should FIP or code be changed?

File: FIPS/fip-0054.md [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+
+_Input parameters_
+
+```rust
+// DAG-CBOR tuple encoded.
+pub struct DelegateCallParams {
+    /// CID of the EVM bytecode to invoke under self's context.
+    pub code: Cid,
+    /// The contract invocation parameters
+    pub input: Vec<u8>, // encoded as a DAG-CBOR byte string
```
Comment (maciejwitowski) [inline FIPS/fip-0054.md:266]: FYI there has been a `caller` field added to this struct https://github.com/filecoin-project/builtin-actors/blob/next/actors/evm/src/lib.rs#L353

File: FIPS/fip-0054.md [hunk 5]
```diff
@@ -0,0 +1,1151 @@
...
+
+**Lifecycle: `CREATE`.**
+Calls the `Create` method of the Ethereum Address Manager ([FIP-0055]) to deploy the smart contract.
+
+**Lifecycle: `CREATE2`.**
+Calls the `Create2` method of the Ethereum Address Manager ([FIP-0055]) to deploy the smart contract.
+
+**Lifecycle: `SELFDESTRUCT`.**
+Marks the current contract as deleted by setting the Tombstone in state.
+Further calls
```
Comment (raulk) [inline FIPS/fip-0054.md:538]: Self-review: dangling sentence.

File: FIPS/fip-0054.md [hunk 6]
```diff
@@ -0,0 +1,1151 @@
...
+We introduce a new built-in actor: the Filecoin EVM (FEVM) runtime actor, capable of running EVM smart contracts on top of the Filecoin Virtual Machine.
+We also introduce various changes to the Filecoin Virtual Machine and to client implementations, necessary to support the operation of this new actor.
+
+## Abstract
+
+The Filecoin EVM (FEVM) runtime built-in actor runs EVM smart contracts compatible with the [Ethereum Paris fork].
+
+It achieves this by embedding a special-purpose EVM interpreter, implementing the integration logic with the Filecoin environment, translating environment-dependent opcodes into their corresponding Filecoin primitives, and mapping all state I/O to the underlying IPLD model.
+
+The EVM interpreter strives for maximal as-is portability. To such avail, it The EVM interpreter supports all required opcodes and Ethereum precompiles to strive for maximal portability.
```
Comment (kaitlin-beegle) [inline FIPS/fip-0054.md:89]: Typo; remove additional "it" after "To such avail,..."

File: FIPS/fip-0054.md [hunk 7]
```diff
@@ -0,0 +1,1151 @@
...
+- Bitwise operations: `BYTE`, `SHL`, `SHR`, `SAR`.
+- Control flow: `STOP`, `JUMPDEST`, `JUMP`, `JUMPI`, `PC`, `STOP`, `RETURN`, `INVALID`.
+- Parameters: `CALLDATALOAD`, `CALLDATASIZE`, `CALLDATACOPY`, `RETURNDATASIZE`, `RETURNDATACOPY`.
+- Stack manipulation: `POP`, `PUSH{1..32}`, `DUP{1..32}`, `SWAP{1..16}`.
+
+#### Opcodes with remarks
+
+**Memory: `MLOAD`, `MSTORE`, `MSTORE8`, `MSIZE`.**
+EVM memory is modelled as an object inside the interpreter, ultimately backed by Wasm memory.
+Usage of of these instructions incurs in Wasm memory expansion costs as per FIP-TODO.
```
Comment (kaitlin-beegle) [inline FIPS/fip-0054.md:447]: Is this reference to an existing FIP draft that isn't yet numbered, or in reference to one which has not yet been drafted? Can we specify?

File: FIPS/fip-0054.md [hunk 8]
```diff
@@ -0,0 +1,1151 @@
...
+   providers to send system messages to continue operating and managing the network.
+   It is hard or impossible to forecast how the gas market will be impacted by the arrival of user-defined contracts to Filecoin.
+   The FVM team has requested the Cryptoecon Lab to produce a community report to evaluate possible scenarios for preparedness.
+
+## Incentive Considerations
+
+This is a technical FIP that independently raises no incentive considerations.
+However, smart contracts may introduce new incentive structures to the Filecoin network that users could engage with.
+
+## Product Considerations
```
Comment (kaitlin-beegle) [inline FIPS/fip-0054.md:1093]: Flagging that details are still needed, or the section ought to be removed.

**Review by kaitlin-beegle — APPROVED** [review]
Reviewed for completeness, accuracy, and language.

File: FIPS/fip-0054.md [hunk 9]
```diff
@@ -0,0 +1,1153 @@
...
+/// # Errors
+///
+/// None
+pub fn context() -> Result<NetworkContext>;
+```
+
+**`network::tipset_cid`**
+
+```rust
+/// Retrieves a tipset's CID within the last finality, if available
```
Comment (anorth) [inline FIPS/fip-0054.md:763]: Nit: suggest a "see below" for this and other new terms that are referenced before being defined.

File: FIPS/fip-0054.md [hunk 10]
```diff
@@ -0,0 +1,1153 @@
...
+```rust
+/// Returns the amount of gas remaining.
+pub fn available() -> Result<u64>;
+```
+
+**`vm::exit`**
+
+```rust
+/// Abort execution with the given code and optional message and data for the return value.
+/// The code and return value are recorded in the receipt, the message is for debugging only.
```
Comment (anorth) [inline FIPS/fip-0054.md:899]: This change in behaviour for all native actors is worth highlighting in discussion.

**Review by anorth — COMMENTED** [review]
Thanks. I have a big discussion item to raise about the privileged treatment of built-in account actors, and assuming they are similar to Ethereum account actors. I will await resolution of which discussion thread to post this (I'm not sure #244 is appropriate).

File: FIPS/fip-0054.md [hunk 11]
```diff
@@ -0,0 +1,1153 @@
+---
+fip: "0054"
+title: Filecoin EVM runtime (FEVM)
+author: Raúl Kripalani (@raulk), Steven Allen (@stebalien)
+discussions-to: https://github.com/filecoin-project/FIPs/discussions/244
```
Comment (jennijuju) [inline FIPS/fip-0054.md:5]:
```suggestion
discussions-to: https://github.com/filecoin-project/FIPs/discussions/592
```

Created a dedicated FEVM discussion per @anorth 's suggestion

File: FIPS/fip-0054.md [hunk 12]
```diff
@@ -0,0 +1,1151 @@
...
+
+It is only be invocable by the Ethereum Address Manager (EAM), specified in [FIP-0055], indirectly via the Init actor.
+This precondition is validated by checking that an f4 address with namespace `10` (corresponding to the EAM) was assigned to the current actor, prior to its `Constructor` method being called.
+
+The Constructor runs the init bytecode with the EVM interpreter, and populates the actor's state accordingly:
+
+- `bytecode` field: the execution bytecode is stored as a raw IPLD block and linked by CID from this field.
+- `contract_state` field: contains the root CID of the storage [KAMT](#kamt-specification) resulting from `SSTORE` operations during construction.
+- `nonce` field: set to 0, unless any calls to CREATE or CREATE2 happen during initialization.
+- `bytecode_hash`: the Keccak-256 hash of the bytecode.
```
[the remark this answers is not in the document]
Comment (mriise) [reply FIPS/fip-0054.md:220]: This is now no longer the case, we switched to using raw 32 byte keccak hashes last week.

File: FIPS/fip-0054.md [hunk 13]
```diff
@@ -0,0 +1,1112 @@
...
+
+Returns the keccak-256 hash of the execution bytecode, which was computed at construction time and stored in state.
+
+_Input parameters_
+
+None.
+
+_Return value_
+
+CID of the contract's bytecode as stored in state.
```
[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md:351]: Adjusted, although we have moved away from a multihash into the actual hash; this was copy-paste.

File: FIPS/fip-0054.md [hunk 14]
```diff
@@ -0,0 +1,1151 @@
...
+- `USR_UNSPECIFIED` (exit code 23) on internal errors instantiating the EVM runtime.
+- `USR_NOT_FOUND` (exit code 17) when failing to load the bytecode.
+- `EVM_CONTRACT_REVERTED` (exit code 33) when the contract reverts.
+- `EVM_CONTRACT_EXECUTION_ERROR` (exit code 34) when the unsupported opcode `CALLCODE` is used.
+- Any other error that is raised during bytecode execution.
+
+#### `GetBytecode` (method number 3)
+
+Returns the CID of the contract's EVM bytecode block, adding it to the caller's reachable set.
+This method is used internally to resolve `EXTCODE*` opcodes.
```
[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md:273]: This sentence helps the reader connect things in their head, I think it's worth keeping. 

> Actual resolution internally is done by looking at the actor's code CID.
I didn't understand this. Could you elaborate?

File: FIPS/fip-0054.md (hunk 7 above)
Comment (raulk) [reply FIPS/fip-0054.md:447]: This was a reference to the newly numbered FIP-0057. Will add that number now -- thanks!"

File: FIPS/fip-0054.md [hunk 15]
```diff
@@ -0,0 +1,1151 @@
...
+There are two kinds of precompiles available to EVM smart contracts running within the FEVM runtime actor:
+
+- Ethereum precompiles
+- Filecoin precompiles
+
+#### Ethereum precompiles
+
+The FEVM runtime actor supports all [Ethereum precompiles] available in the Ethereum Paris fork.
+These precompiles sit at their original Ethereum addresses.
+There are no functional departures with respect to their original behavior.
```
[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md]: Thanks, I don't think that particular detail is worth mentioning. Just explaining that precompiles don't do any Ethereum gas accounting and are charged per execution cost/syscalls performed is enough.

File: FIPS/fip-0054.md [hunk 16]
```diff
@@ -0,0 +1,1153 @@
...
+Returns the base fee plus the gas premium, in atto precision (same precision as Ethereum).
+This is functionally equivalent to Ethereum's `GASPRICE` definition following [EIP-1559], which is to return the `effective_gas_price`, defined as:
+
+```
+effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas
+```
+
+**Accessors: `BLOCKHASH`.**
+Returns hash payload extracted from the multihash inlined in the tipset CID of the requested epoch, with a maximum lookback of 256 epochs, truncated to preserve the left-most 32 bytes.
+TODO: null rounds.
```
[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md]: Fixed.

File: FIPS/fip-0054.md [hunk 17]
```diff
@@ -0,0 +1,1153 @@
...
+**Lifecycle: `CREATE2`.**
+Calls the `Create2` method of the Ethereum Address Manager ([FIP-0055]) to deploy the smart contract.
+
+**Lifecycle: `SELFDESTRUCT`.**
+Marks the current contract as deleted by setting the Tombstone in state.
+Further calls
+
+**Calls: `CALL`.**
+Performs a call to another actor, behaving differently depending on the target's actor type.
+If the target is another EVM smart contract or a non-account Wasm actor, it invokes its `InvokeContract` method (2).
```
[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md]: Indeed, matching against the actor type for the code CID of the callee (`actor::get_builtin_actor_type` and `actor::get_actor_code_cid` syscalls).

[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md]: Discussion continued in https://github.com/filecoin-project/FIPs/discussions/592#discussioncomment-4700971.

[the remark this answers is not in the document]
Comment (raulk) [reply FIPS/fip-0054.md]: I will specify how we do it today, but let's continue discussing potential changes in that thread?

File: FIPS/fip-0054.md (hunk 9 above)
Comment (raulk) [reply FIPS/fip-0054.md:862]: Added.

File: FIPS/fip-0054.md (hunk 10 above)
Comment (raulk) [reply FIPS/fip-0054.md:1001]: Done.

File: FIPS/fip-0054.md [hunk 18]
```diff
@@ -0,0 +1,1296 @@
...
+
+```rust
+/// DAG-CBOR tuple-encoded.
+pub struct State {
+    /// The EVM contract bytecode resulting from calling the
+    /// initialization code (init code) by the constructor.
+    pub bytecode: Cid,
+
+    /// The EVM contract bytecode hash keccak256(bytecode)
+    pub bytecode_hash: multihash::Multihash,
```
Comment (Stebalien) [inline FIPS/fip-0054.md:144]: Changed to a 20 byte array (no multihash prefix).

File: FIPS/fip-0054.md [hunk 19]
```diff
@@ -0,0 +1,1296 @@
...
+/// Gets the CID for a given tipset.
+///
+/// If the supplied epoch happens to be a null round, the client must return the CID
+/// of the previous first non-null round.
+fn get_tipset_cid(&self, epoch: ChainEpoch) -> anyhow::Result<Cid>;
+```
+
+### New general exit codes
+
+- USR_READ_ONLY (exit code 25). To be returned by actors when the requested operation cannot be performed in "read-only" mode.
```
Comment (Stebalien) [inline FIPS/fip-0054.md:1089]: Applies to:

1. Value transfers.
2. `set_root`
3. `selfdestruct`.
4. Actor creation.

File: FIPS/fip-0054.md [hunk 20]
```diff
@@ -0,0 +1,1312 @@
...
+See remarks on the `DELEGATECALL` opcode below for more information.
+
+_Input parameters_
+
+```rust
+// DAG-CBOR tuple encoded.
+pub struct DelegateCallParams {
+    pub code: Cid,
+    /// The contract invocation parameters
+    /// Encoded as a DAG-CBOR byte string
```
Comment (vyzo) [inline FIPS/fip-0054.md:346]: plain CBOR?

File: FIPS/fip-0054.md [hunk 21]
```diff
@@ -0,0 +1,1312 @@
...
+
+_Errors_
+
+- Exit code `USR_FORBIDDEN` (18) when caller is not self.
+
+#### `InvokeContract` (method number 3844450837[^1])
+
+Invokes an EVM smart contract by loading the execution bytecode from state, and dispatching input data to it.
+
+The input data is expected to be framed as a DAG-CBOR byte string.
```
Comment (vyzo) [inline FIPS/fip-0054.md:367]: again, isnt this plain CBOR?

File: FIPS/fip-0054.md [hunk 22]
```diff
@@ -0,0 +1,1312 @@
...
+- Boolean operators: `LT`, `GT`, `SLT`, `SGT`, `EQ`, `ISZERO`, `AND`, `OR`, `XOR`, `NOT`.
+- Bitwise operations: `BYTE`, `SHL`, `SHR`, `SAR`.
+- Control flow: `STOP`, `JUMPDEST`, `JUMP`, `JUMPI`, `PC`, `STOP`, `RETURN`, `INVALID`.
+- Parameters: `CALLDATALOAD`, `CALLDATASIZE`, `CALLDATACOPY`, `RETURNDATASIZE`, `RETURNDATACOPY`.
+- Stack manipulation: `POP`, `PUSH{0..32}`, `DUP{1..32}`, `SWAP{1..16}`.
+
+#### Opcodes with remarks
+
+**Memory: `MLOAD`, `MSTORE`, `MSTORE8`, `MSIZE`.**
+EVM memory is modelled as an object inside the interpreter, ultimately backed by Wasm memory.
```
Comment (vyzo) [inline FIPS/fip-0054.md:463]: We should probably specify that we enforce a 32bit limit?
This has different failure semantics than ETH which will run out of gas.

File: FIPS/fip-0054.md [hunk 23]
```diff
@@ -0,0 +1,1312 @@
...
+
+**Accessors: `CALLVALUE`.**
+Returns the filecoin value sent in the message, in atto precision (same precision as Ethereum).
+
+**Accessors: `GASPRICE`.**
+Returns the base fee plus the gas premium, in atto precision (same precision as Ethereum).
+This is functionally equivalent to Ethereum's `GASPRICE` definition following [EIP-1559], which is to return the `effective_gas_price`, defined as:
+
+```
+effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas
```
Comment (vyzo) [inline FIPS/fip-0054.md:486]: We have capped this.

File: FIPS/fip-0054.md [hunk 24]
```diff
@@ -0,0 +1,1312 @@
...
+Returns the base fee, in atto precision (same precision as Ethereum).
+
+**Accessors: `SELFBALANCE`.**
+Returns the contract's filecoin balance, in atto precision (same precision as Ethereum).
+
+**Accessors: `GAS`.**
+Returns the Filecoin gas available.
+
+**Control flow: `REVERT`.**
+Calls the `fvm::exit` syscall with user-space exit code 33 (`EVM_CONTRACT_REVERTED`), passing in the supplied value and reverting all changes.
```
Comment (vyzo) [inline FIPS/fip-0054.md:526]: It is more nuanced than that, we should probably say it aborts execution.

File: FIPS/fip-0054.md [hunk 25]
```diff
@@ -0,0 +1,1312 @@
...
+///
+/// None
+pub fn message_context() -> Result<MessageContext>;
+```
+
+### Changed syscalls
+
+#### `send::send` syscall
+
+This syscall now takes an extra `send_flags` u64 field acting as a bitmap. The possible flags are:
```
Comment (vyzo) [inline FIPS/fip-0054.md:1060]: Also gas?

File: FIPS/fip-0054.md [hunk 26]
```diff
@@ -0,0 +1,1312 @@
...
+Upgrades will impact gas costs in hard-to-predict ways, with compute gas being the most sensitive component.
+We discourage smart contract developers to rely on specific gas values in their contract logic.
+This includes gas limits passed in `*CALL*` operations.
+While the system honors those limits, costs will change across upgrades, so it's unsafe to hardcode assumptions.
+
+### Main differences between Filecoin's EVM and Ethereum's EVM
+
+1. Gas metering and execution halt are performed according to the Filecoin gas model.
+   Ethereum gas accounting does not apply in the Filecoin EVM.
+2. Gas limits set when CALLing precompiles do not apply.
```
Comment (vyzo) [inline FIPS/fip-0054.md:1242]: Is this accurate for call_actor?

File: FIPS/fip-0054.md [hunk 27]
```diff
@@ -0,0 +1,1312 @@
...
+3. Filecoin gas costs are not stable over network updates.
+   Developers must not make gas usage assumptions, as Filecoin gas is highly sensitive to implementation details, and they should refrain from using hardcoded gas limits in calls.
+4. Bare value sends (messages with method number 0) to Filecoin EVM smart contracts do not trigger smart contract logic, even if the message carries parameters.
+5. The `CALLCODE` opcode is not supported.
```
Comment (vyzo) [inline FIPS/fip-0054.md:1246]: Note that this is deprecated anyway.

Comment (raulk) [reply FIPS/fip-0054.md:1246]: Technically it is not. DELEGATECALL was introduced and Solidity removed callcode, but it still exists in the Ethereum protocol. But I can add a note.

File: FIPS/fip-0054.md (hunk 25 above)
Comment (raulk) [reply FIPS/fip-0054.md:1060]: `gas_limit`, good catch.

File: FIPS/fip-0054.md (hunk 23 above)
Comment (raulk) [reply FIPS/fip-0054.md:486]: Yep, this was a recent change. https://github.com/filecoin-project/ref-fvm/pull/1512

File: FIPS/fip-0054.md (hunk 24 above)
Comment (raulk) [reply FIPS/fip-0054.md:526]: An exit with a non-zero exit code implies an abort.

File: FIPS/fip-0054.md (hunk 22 above)
Comment (raulk) [reply FIPS/fip-0054.md:463]: Done.

File: FIPS/fip-0054.md (hunk 21 above)
Comment (raulk) [reply FIPS/fip-0054.md:367]: It is as of https://github.com/filecoin-project/ref-fvm/pull/1509.

File: FIPS/fip-0054.md (hunk 20 above)
Comment (raulk) [reply FIPS/fip-0054.md:346]: It is as of https://github.com/filecoin-project/ref-fvm/pull/1509.

File: FIPS/fip-0054.md (hunk 24 above)
Comment (raulk) [reply FIPS/fip-0054.md:526]: Clarified.

File: FIPS/fip-0054.md (hunk 26 above)
Comment (raulk) [reply FIPS/fip-0054.md:1242]: Good catch.

File: FIPS/fip-0054.md (hunk 27 above)
Comment (raulk) [reply FIPS/fip-0054.md]: Addressed.

**Pull request merged**